server/src/main/java/io/druid/server/namespace/NamespacedExtractionModule.java

```diff
@@ -0,0 +1,448 @@
+import java.util.concurrent.Executors;
+
+/**
+ *
+ */
+public class NamespacedExtractionModule implements DruidModule
+{
+  private static final String NAMESPACE_ENABLED_KEY = "enabledNamespaces";
+  private static final Logger log = new Logger(NamespacedExtractionModule.class);
+  private static final String TYPE_PREFIX = "druid.query.extraction.namespace.cache";
+  private final ConcurrentMap<String, Function<String, String>> fnCache = new ConcurrentHashMap<>();
+
+  @Override
+  public List<? extends Module> getJacksonModules()
+  {
+    return ImmutableList.<Module>of(
+        new Module()
+        {
+          @Override
+          public String getModuleName()
+          {
+            return "DruidNamespacedExtractionModule";
+          }
+
+          @Override
+          public Version version()
+          {
+            return Version.unknownVersion();
+          }
+
+          @Override
+          public void setupModule(SetupContext context)
+          {
+            context.registerSubtypes(ExtractionNamespace.class);
+          }
+        }
+    );
+  }
+
+  @Override
+  public void configure(Binder binder)
+  {
+    JsonConfigProvider.bind(binder, TYPE_PREFIX, NamespaceExtractionCacheManager.class);
+
+    binder
+        .bind(ExtractionNamespaceFunctionFactory.class)
+        .annotatedWith(Names.named(JDBCExtractionNamespace.class.getCanonicalName()))
+        .to(JDBCExtractionNamespaceFunctionFactory.class)
+        .in(LazySingleton.class);
+    binder
+        .bind(ExtractionNamespaceFunctionFactory.class)
+        .annotatedWith(Names.named(URIExtractionNamespace.class.getCanonicalName()))
+        .to(URIExtractionNamespaceFunctionFactory.class)
+        .in(LazySingleton.class);
+
+    binder
+        .bind(Key.get(Object.class, Names.named(NAMESPACE_ENABLED_KEY)))
+        .toInstance(new Object());
+
+    LifecycleModule.register(binder, NamespacedKeeper.class);
+  }
+
+  public static boolean isNamespaceEnabled(Injector injector)
+  {
+    return null != injector.getExistingBinding(Key.get(Object.class, Names.named(NAMESPACE_ENABLED_KEY)));
+  }
+
+  @Provides
+  @Named("namespaceExtractionFunctionCache")
+  public ConcurrentMap<String, Function<String, String>> getFnCache()
+  {
+    return fnCache;
+  }
+
+  @Provides
+  @Named("dimExtractionNamespace")
+  @LazySingleton
+  public Function<String, Function<String, String>> getFunctionMaker(
+      @Named("namespaceExtractionFunctionCache")
+      final ConcurrentMap<String, Function<String, String>> fnCache
+  )
+  {
+    return new Function<String, Function<String, String>>()
+    {
+      @Nullable
+      @Override
+      public Function<String, String> apply(final String namespace)
+      {
+        Function<String, String> fn = fnCache.get(namespace);
+        if (fn == null) {
+          throw new IAE("Namespace [%s] not found", namespace);
+        }
+        return fn;
+      }
+    };
+  }
+
+  @ManageLifecycle
+  public static class NamespacedKeeper
```

Comment:
It's rather weird that this class is in a Module, I think.  I also am having a hard time understanding the semantic function of the class.  Please add some javadoc and consider making this a first-class class.